File: Lecture 2 - Arrays/Problem Set 2/Readability/readability.c
/*
    READABILITY PROGRAM (STANDARD C VERSION)

    Problem:
    Calculate the readability grade level of text using
    the Coleman-Liau Index formula.

    No CS50 library used.
*/

#include <ctype.h>
#include <math.h>
#include <stdio.h>
#include <string.h>

// Function prototypes
int letters_count(const char *text);
int word_count(const char *text);
int sentence_count(const char *text);
int Coleman_Liau_index(int letters, int words, int sentences);

int main(void)
{
    char text[1000];

    printf("Text: ");
    fgets(text, sizeof(text), stdin);

    int letters = letters_count(text);
    int words = word_count(text);
    int sentences = sentence_count(text);

    int index = Coleman_Liau_index(letters, words, sentences);

    if (index < 1)
    {
        printf("Before Grade 1\n");
    }
    else if (index > 16)
    {
        printf("Grade 16+\n");
    }
    else
    {
        printf("Grade %d\n", index);
    }

    return 0;
}

int letters_count(const char *text)
{
    int letters = 0;

    for (int i = 0; text[i] != '\0'; i++)
    {
        if (isalpha(text[i]))
        {
            letters++;
        }
    }

    return letters;
}

int word_count(const char *text)
{
    int spaces = 0;

    for (int i = 0; text[i] != '\0'; i++)
    {
        if (isblank(text[i]))
        {
            spaces++;
        }
    }

    return spaces + 1;
}

int sentence_count(const char *text)
{
    int sentences = 0;

    for (int i = 0; text[i] != '\0'; i++)
    {
        if (text[i] == '.' || text[i] == '!' || text[i] == '?')
        {
            sentences++;
        }
    }

    return sentences;
}

int Coleman_Liau_index(int letters, int words, int sentences)
{
    if (words == 0)
    {
        return 0;
    }

    float L = ((float) letters / words) * 100.0;
    float S = ((float) sentences / words) * 100.0;

    float index = (0.0588 * L) - (0.296 * S) - 15.8;

    return (int) round(index);
}

/*
---------------- ALGORITHM ----------------

1. Read input text using fgets().
2. Count letters using isalpha().
3. Count words by counting spaces and adding 1.
4. Count sentences by checking '.', '!', '?'.
5. Compute averages:
       L = letters per 100 words
       S = sentences per 100 words
6. Apply Coleman-Liau formula.
7. Round result.
8. Print grade level accordingly.
9. End.

Time Complexity: O(n)
Space Complexity: O(1)

Note:
fgets() includes newline character '\n',
but it does not affect counting logic.
*/

File: Lecture 6 - Python/Problem Set 6/Readability/readability.py
"""
PROBLEM: READABILITY (COLEMAN–LIAU INDEX)

Write a program that:
1. Prompts the user for text.
2. Counts:
   - Letters
   - Words
   - Sentences
3. Computes Coleman–Liau index.
4. Prints grade level.

Formula:
index = 0.0588L - 0.296S - 15.8

Where:
L = average number of letters per 100 words
S = average number of sentences per 100 words
"""


def letters_count(text):
    """Count alphabetic characters."""
    letters = 0
    for char in text:
        if char.isalpha():
            letters += 1
    return letters


def word_count(text):
    """Count words based on spaces."""
    spaces = 0
    for char in text:
        if char.isspace():
            spaces += 1
    return spaces + 1


def sentence_count(text):
    """Count sentences based on '.', '!', '?'."""
    sentences = 0
    for char in text:
        if char in ".!?":
            sentences += 1
    return sentences


def coleman_liau_index(letters, words, sentences):
    """Compute readability index."""
    if words == 0:
        return 0

    L = letters / words * 100
    S = sentences / words * 100

    index = 0.0588 * L - 0.296 * S - 15.8
    return round(index)


def main():
    text = input("Text: ")

    letters = letters_count(text)
    words = word_count(text)
    sentences = sentence_count(text)

    index = coleman_liau_index(letters, words, sentences)

    if index < 1:
        print("Before Grade 1")
    elif index > 16:
        print("Grade 16+")
    else:
        print(f"Grade {index}")


if __name__ == "__main__":
    main()


"""
---------------- ALGORITHM ----------------

1. Take text input from user.
2. Count letters using isalpha().
3. Count words by counting spaces + 1.
4. Count sentences by checking '.', '!', '?'.
5. Compute:
      L = letters / words * 100
      S = sentences / words * 100
6. Apply Coleman–Liau formula.
7. Round result.
8. Print grade level.

Time Complexity: O(n)
Space Complexity: O(1)
"""